urlcdab7d8364_jrwalsh_CycTools_tgz-pJ8-edu_iastate_cyctools_MainJ8/src/edu/iastate/cyctools/externalSourceCode/DiffMatchPatch.java

```diff
@@ -0,0 +1,2535 @@
+    if (Diff_Timeout <= 0) {
+      deadline = Long.MAX_VALUE;
+    } else {
+      deadline = System.currentTimeMillis() + (long) (Diff_Timeout * 1000);
+    }
+    return diff_main(text1, text2, checklines, deadline);
+  }
+
+  /**
+   * Find the differences between two texts.  Simplifies the problem by
+   * stripping any common prefix or suffix off the texts before diffing.
+   * @param text1 Old string to be diffed.
+   * @param text2 New string to be diffed.
+   * @param checklines Speedup flag.  If false, then don't run a
+   *     line-level diff first to identify the changed areas.
+   *     If true, then run a faster slightly less optimal diff.
+   * @param deadline Time when the diff should be complete by.  Used
+   *     internally for recursive calls.  Users should set DiffTimeout instead.
+   * @return Linked List of Diff objects.
+   */
+  private LinkedList<Diff> diff_main(String text1, String text2,
+                                     boolean checklines, long deadline) {
+    // Check for null inputs.
+    if (text1 == null || text2 == null) {
+      throw new IllegalArgumentException("Null inputs. (diff_main)");
+    }
+
+    // Check for equality (speedup).
+    LinkedList<Diff> diffs;
+    if (text1.equals(text2)) {
+      diffs = new LinkedList<Diff>();
+      if (text1.length() != 0) {
+        diffs.add(new Diff(Operation.EQUAL, text1));
+      }
+      return diffs;
+    }
+
+    // Trim off common prefix (speedup).
+    int commonlength = diff_commonPrefix(text1, text2);
+    String commonprefix = text1.substring(0, commonlength);
+    text1 = text1.substring(commonlength);
+    text2 = text2.substring(commonlength);
+
+    // Trim off common suffix (speedup).
+    commonlength = diff_commonSuffix(text1, text2);
+    String commonsuffix = text1.substring(text1.length() - commonlength);
+    text1 = text1.substring(0, text1.length() - commonlength);
+    text2 = text2.substring(0, text2.length() - commonlength);
+
+    // Compute the diff on the middle block.
+    diffs = diff_compute(text1, text2, checklines, deadline);
+
+    // Restore the prefix and suffix.
+    if (commonprefix.length() != 0) {
+      diffs.addFirst(new Diff(Operation.EQUAL, commonprefix));
+    }
+    if (commonsuffix.length() != 0) {
+      diffs.addLast(new Diff(Operation.EQUAL, commonsuffix));
+    }
+
+    diff_cleanupMerge(diffs);
+    return diffs;
+  }
+
+  /**
+   * Find the differences between two texts.  Assumes that the texts do not
+   * have any common prefix or suffix.
+   * @param text1 Old string to be diffed.
+   * @param text2 New string to be diffed.
+   * @param checklines Speedup flag.  If false, then don't run a
+   *     line-level diff first to identify the changed areas.
+   *     If true, then run a faster slightly less optimal diff.
+   * @param deadline Time when the diff should be complete by.
+   * @return Linked List of Diff objects.
+   */
+  private LinkedList<Diff> diff_compute(String text1, String text2,
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 12. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 258-276 into a separate method.